docs/docs/reconciliation.md

```diff
@@ -0,0 +1,123 @@
+
+## Motivation
+
+When you use React, at a single point in time you can think of the `render()` function as creating a tree of React elements. On the next state or props update, that `render()` function will return a different tree of React elements. The React framework then needs to figure out a way to efficiently convert the first tree into the second tree.
+
+There are some generic solutions to this algorithmic problem of generating the minimum number of operations to transform one tree into another. However, the [state of the art algorithms](http://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) have a complexity in the order of O(n<sup>3</sup>) where n is the number of elements in the tree.
+
+If we used this in React, displaying 1000 elements would require in the order of one billion comparisons. This is far too expensive. Instead, React implements a heuristic O(n) algorithm based on two assumptions:
+
+1. Two elements of different types can be treated as totally different trees.
+2. The developer can hint at which child elements may be stable across different renders with a `key` prop.
+
+In practice, these assumptions are valid for almost all practical use cases.
+
+## The Diffing Algorithm
+
+When diffing two trees, React first compares the types of the two root elements. If the root elements have different types, React just throws away the first tree and builds the second tree from scratch. When the root elements have the same type, React first converts the root element, and then recurses. The specifics are different depending on whether the elements are DOM elements or component elements.
+
+### DOM Elements
+
+When comparing two React DOM elements, React looks at the attributes of both, keeps the same underlying DOM element, and only updates the changed attributes. For example:
+
+```xml
+<div className="before" title="stuff" />
+
+<div className="after" title="stuff" />
+```
+
+When converting between these two elements, React knows to only modify the `className`.
+
+When updating `style`, React also knows to update only the properties that changed. For example:
+
+```xml
+<div style={{'{{'}}color: 'red', fontWeight: 'bold'}} />
+
+<div style={{'{{'}}color: 'green', fontWeight: 'bold'}} />
+```
+
+When converting between these two elements, React knows to only modify the `color` style, not the `fontWeight`.
+
+After handling the DOM element, React then recurses on the children.
+
+### Custom Components
+
+When a component updates, the instance stays the same, so that state is maintained across renders. React takes all the attributes from the new component element and calls `componentWillReceiveProps()` and `componentWillUpdate()` on the previous one.
```

Comment:
Similarly, let's first add a section that describes what happens when types are different. The old one receives `componentWillUnmount()` and the whole tree is destroyed, the next one receives `componentWillMount()` and `componentDidMount()`. The state associated with the old one gets lost.

Then explain what happens during an update.

>React takes all the attributes from the new component element and calls `componentWillReceiveProps()` and `componentWillUpdate()` on the previous one.

This is incorrect, lifecycle methods don't live on an element. Element is just a plain object description (`<Foo />` or `{ type: Foo }`), there are no methods on it. React doesn't call any methods on elements.

If you define component as a class, React internally maintains its instance. This is the one that has all the methods and has `this.props`. Functional components don't have instances.

When React sees that `App` component returned `<Foo prop={42} />` (or `{ type: Foo, props: { prop: 42 } }`) from `render()` the first time, it creates an instance of `Foo` and assigns its `this.props` by reading them from the `<Foo prop={42} />` element.

If `App` component returns `<Foo prop={43} />` the next time, React will know that a `Foo` instance already exists, corresponding to the previous element "on the same spot". So instead of recreating the instance, it will update its `this.props` to point to new element's props (`{ prop: 42 }`), and call `componentWillReceiveProps()` on it.

Elements and instances are different things. Elements describe the UI. They are short-lived and immutable. Instances are things that React creates to represent "stable" pieces of UI and hold state associated with them. Over its lifetime, and instance can "receive" several elements as it gets updated. React will not destroy the instance as long as the parent component keeps returning an element with the same type (and key) at the same spot.

(I don't ask that you write all of this, but I hope this helps figuring out how to phrase it correctly.)